calyx-opt/src/passes/metadata_table_gen.rs

```diff
@@ -0,0 +1,110 @@
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir::Id;
+use calyx_utils::WithPos;
+use linked_hash_map::LinkedHashMap;
+use std::fmt;
+use std::path::PathBuf;
+
+/// Metadata stores a Map between each group name and data used in the metadata table (specified in PR #2022)
+#[derive(Default)]
+pub struct Metadata {
+    groups: LinkedHashMap<Id, (usize, PathBuf)>,
+}
+
+impl Metadata {
+    /// Create an empty metadata table
+    pub fn new() -> Self {
+        Metadata {
+            groups: LinkedHashMap::new(),
+        }
+    }
+
+    /// Add a new entry to the metadata table
+    fn add_entry(&mut self, name: Id, line: usize, path: PathBuf) -> &mut Self {
+        let ins = self.groups.insert(name, (line, path));
+        match ins {
+            None => self,
+            Some(_v) => {
+                println!("Found two of same group name");
```

Comment:
Let's just have this panic for the time being if we have two groups of the same
name since we need to modify the format going forward to account for this